### PR TITLE
Support complex macros.

### PR DESCRIPTION
**Pull request by reitermarkus**
This supersedes https://github.com/rust-lang/rust-bindgen/pull/2268 which aimed to support casts in macros.

The [`cmacro`](https://github.com/reitermarkus/cmacro-rs) crate I developed over the last months allows parsing casts in variable-like macros and even more complex macros, such as function-like macros and nested macros.

Macro parsing has been split from `Var` into a separate `MacroDef` type, since macros need to be parsed last. Variables can be evaluated at parse time, macros on the other hand can only be fully evaluated after all types, variables, functions and other macros have been parsed. This allows getting the correct final value and is needed for parsing nested macros.

### CODE REVIEW

**Comment by reitermarkus** [comment]
Can someone please take a look at this PR? It's been over three months without a review.

**Comment by pvdrz** [comment]
:wave: Hi! Sorry for not giving you an answer in all this time. Somehow I missed this and wasn't aware of it.

I will properly review this PR as I think it might solve many issues regarding the type that is assigned to these define macros when they get translated as constants. Bindgen has depended for a long time on `rust-cexpr` but having someone breathe new life to this part of bindgen is good IMHO.

That being said, we still should figure out the logistics of maintenance due to bindgen using your crate and I don't have the whole picture in my head so I'll defer that part to @emilio

**Review by pvdrz — CHANGES_REQUESTED** [review]
I'm done with the review :tada: 

I apologize in advance if I'm being harsh on some comments. I am really interested on understanding how all of this works :grin:

File: .github/workflows/bindgen.yml [hunk 1]
```diff
@@ -2,11 +2,11 @@ name: bindgen
 
 on:
   push:
-    branches:
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply .github/workflows/bindgen.yml:5]: Yes, just temporarily enabled this so CI runs on my fork.

File: bindgen-tests/tests/expectations/tests/macro_const.rs [hunk 2]
```diff
@@ -5,10 +5,16 @@
...
+    const BYTES: [u8; 4] = *b"bar\0";
+    BYTES.as_ptr() as *const ::std::os::raw::c_char
+};
+pub const CHAR: ::std::os::raw::c_char = 'b' as ::std::os::raw::c_char;
+pub const CHARR: ::std::os::raw::c_char = '\0' as ::std::os::raw::c_char;
+pub const FLOAT: f32 = 5.09;
+pub const FLOAT_EXPR: f32 = 0.005;
 pub const LONG: u32 = 3;
-pub const INVALID_UTF8: [u8; 5usize] = [240u8, 40u8, 140u8, 40u8, 0u8];
+pub const INVALID_UTF8: *const ::std::os::raw::c_char = {
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: In C, a string literal is a `const *char`, so in Rust it should be the same.

I guess using `&CStr` will make more sense once all needed functions are `const`-stable.

File: bindgen-tests/tests/headers/layout_array.h [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// bindgen-flags: --with-derive-hash --with-derive-partialeq --with-derive-eq --impl-partialeq --rust-target 1.40
+// bindgen-flags: --with-derive-hash --with-derive-partialeq --with-derive-eq --impl-partialeq --rust-target 1.40 --blocklist-function "__rte_aligned"
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/headers/layout_array.h:1]: This is a C attribute macro, so produced a useless Rust macro.

I'm not sure if I added this to keep the diff smaller or if it produces wrong Rust code.

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/headers/layout_array.h:1]: Yeah, it produces this useless macro which cannot be called since `__attribute__` and `__alligned__` don't exist:

```rust
#[doc(hidden)]
#[macro_export]
macro_rules! __cmacro____rte_aligned {
    ($ a : expr) => {
        __attribute__((__aligned__(($a).into())).into())
    };
}
```

File: bindgen-tests/tests/expectations/tests/macro_const_1_0.rs [hunk 4]
```diff
@@ -5,10 +5,16 @@
...
-pub const CHAR: u8 = 98u8;
-pub const CHARR: u8 = 0u8;
-pub const FLOAT: f64 = 5.09;
-pub const FLOAT_EXPR: f64 = 0.005;
+pub const foo: *const ::std::os::raw::c_char = {
+    const BYTES: [u8; 4] = *b"bar\0";
+    BYTES.as_ptr() as *const ::std::os::raw::c_char
+};
+pub const CHAR: ::std::os::raw::c_char = 'b' as ::std::os::raw::c_char;
+pub const CHARR: ::std::os::raw::c_char = '\0' as ::std::os::raw::c_char;
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const_1_0.rs:13]: Really depends on whether we want `char`s to be `c_char` or `u8` in Rust.

File: bindgen/callbacks.rs [hunk 5]
```diff
@@ -62,7 +62,16 @@ pub trait ParseCallbacks: fmt::Debug {
...
     /// the expansion of the macro as a sequence of tokens.
-    fn func_macro(&self, _name: &str, _value: &[&[u8]]) {}
+    fn func_macro(&self, _name: &str, _args: &[&str], _value: &[&[u8]]) {}
+
+    /// Specify the type of a macro argument.
+    ///
+    /// This is needed if you want to generate a function instead of a macro.
+    /// If all argument types and the return type of a macro can be inferred,
+    /// a function will be generated instead of a macro.
+    fn func_macro_arg_type(&self, _name: &str, _arg: &str) -> Option<String> {
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/callbacks.rs:72]: Say you have a macro like

```c
#define times(x, y) (x * y)
```

This will normally generate

```rust
macro_rules! times {
  ($x:expr, $y:expr) => {{ $x * $y }};
}
```

If you specify the types for each argument with `func_macro_arg_type`, i.e. 

`func_macro_arg_type("times", "x") == Some("int")` 

and

`func_macro_arg_type("times", "y") == Some("int")`

it will generate a function instead, since all types can be resolved:

```rust
fn times(x: c_int, y: c_int) {
  x * y
}
```

File: bindgen-tests/tests/headers/macro_func_wrapper.h [hunk 6]
```diff
@@ -0,0 +1,4 @@
+void func(int, int);
+
+#define Y 7
+#define wrapper_func(x) func(x, Y)
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/headers/macro_func_wrapper.h:4]: Seems like it is currently ignored, probably since macro parsing stops if the macro name is encountered.

File: bindgen-tests/tests/expectations/tests/macro-cast.rs [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals
+)]
+
+pub const ULONG_MAX: u32 = 4294967295;
+pub const portMAX_DELAY: TickType_t = 4294967295u32 as TickType_t;
+pub const NEG_TO_POS: ::std::os::raw::c_uint = -1i8 as ::std::os::raw::c_uint;
+pub const BIG_TO_SMALL: ::std::os::raw::c_ushort =
+    4294967295u32 as ::std::os::raw::c_ushort;
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro-cast.rs:12]: Yes, correct.

File: bindgen/clang.rs [hunk 8]
```diff
@@ -2149,29 +2118,22 @@ impl EvalResult {
...
-    pub(crate) fn as_int(&self) -> Option<i64> {
+    pub(crate) fn as_int(&self) -> Option<i128> {
         if self.kind() != CXEval_Int {
             return None;
         }
 
         if unsafe { clang_EvalResult_isUnsignedInt(self.x) } != 0 {
             let value = unsafe { clang_EvalResult_getAsUnsigned(self.x) };
-            if value > i64::max_value() as c_ulonglong {
+            if value as u128 > i128::MAX as u128 {
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/clang.rs]: Yes, done.

File: bindgen-tests/tests/expectations/tests/macro-do-while.rs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    non_upper_case_globals
+)]
+
+#[allow(non_snake_case, unused_mut)]
+#[inline(always)]
+pub unsafe extern "C" fn vTaskDelayUntil(
+    mut pxPreviousWakeTime: ::std::os::raw::c_int,
+    mut xTimeIncrement: ::std::os::raw::c_int,
+) {
+    loop {
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro-do-while.rs:14]: This is done in `cmacro`.

File: bindgen-tests/tests/headers/macro_func_wrapper.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/headers/macro_func_wrapper.h:4]: First, the macro cannot be generated since it thinks it is a recursive definition:

```
Cannot generate function macro: RecursiveDefinition("func")
```

Next, the `func` function is not generated since the macro previously called `saw_function`.

So apart from the recursive definition issue, we need to differentiate between `saw_function` and `saw_fn_macro`.

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/headers/macro_func_wrapper.h:4]: This edge case is now supported.

File: bindgen-tests/tests/expectations/tests/macro-expr-basic.rs [hunk 10]
```diff
@@ -8,7 +8,7 @@
 pub const FOO: u32 = 1;
 pub const BAR: u32 = 4;
 pub const BAZ: u32 = 5;
-pub const MIN: i64 = -9223372036854775808;
+pub const MIN: u64 = 9223372036854775808;
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro-expr-basic.rs:11]: This changed since previously numbers were cast to `i64` internally in `cexpr`. Now that `i128` is used, it actually fits in the positive part of `i128`, resulting in `u64` being used.

File: bindgen-integration/Cargo.toml [hunk 11]
```diff
@@ -2,6 +2,7 @@
 name = "bindgen-integration"
 description = "A package to test various bindgen features"
 version = "0.1.0"
+edition = "2018"
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-integration/Cargo.toml:5]: Just checked. Exporting macros as items (`pub __cmacro_MACRO as MACRO;`) is not supported without it.

File: bindgen/Cargo.toml [hunk 12]
```diff
@@ -55,3 +55,6 @@ experimental = []
 testing_only_extra_assertions = []
 testing_only_libclang_9 = []
 testing_only_libclang_5 = []
+
+# [patch."https://github.com/reitermarkus/cmacro-rs"]
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/Cargo.toml]: Moved it to the top-level `Cargo.toml` for now and will remove once the PR is ready to merge.

File: bindgen-tests/tests/expectations/tests/macro-expr-basic.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply bindgen-tests/tests/expectations/tests/macro-expr-basic.rs:11]: Oh OK I get it now. So if the original value is positive you try to fit in in a unsigned integer.

File: bindgen-tests/tests/expectations/tests/macro_const.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: Hmm I checking the standard and it says
> In translation phase 7, a byte or code of value zero is appended to each multibyte character sequence that results from a string literal or literals. The multibyte character sequence is then used to initialize an array of static storage duration and length just sufficient to contain the sequence.

I understand that for any evaluation such as `#if STR_LIT_1 == 0` , the literal would get converted into a pointer but from the standard I would understand that they are treated as arrays first. What do you think?

File: bindgen/callbacks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply bindgen/callbacks.rs:72]: Hmmm interesting. What happens if you use both callbacks with the same macro?

**Review by pvdrz — CHANGES_REQUESTED** [review]
I added a couple extra comments

Also a lot of people want function-like macros just to be able to use simple things like object-like macros that use `_IO` from `<ioctl.h>`

I'm pretty sure it already works but could you add a test for something like

```h
#define _IO(a_) (a_)

#define UI_DEV_CREATE _IO(1)
#define UI_DEV_DESTROY _IO(2)
```

In that way we could also close https://github.com/rust-lang/rust-bindgen/issues/753

File: bindgen/callbacks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/callbacks.rs:72]: You mean both `func_macro` and `func_macro_arg_type`?

Nothing will happen. `func_macro` let's the user do whatever they want with the macro as input, it doesn't have any effect on what is generated. `func_macro` will probably not be needed as much since many macros which previously needed some custom generation are supported now.

**Comment by pvdrz** [comment]
Also random question. Is token concatenation using `##` supported?

**Comment by reitermarkus** [comment]
Yes, concatenation (`ident ## ident`) as well as stringification (`#ident`) are supported. I think the only problem is Rust's `concat_idents!` does not (yet) support concatenating numbers (i.e. `ident ## 1`).

**Comment by pvdrz** [comment]
> Yes, concatenation (`ident ## ident`) as well as stringification (`#ident`) are supported. I think the only problem is Rust's `concat_idents!` does not (yet) support concatenating numbers (i.e. `ident ## 1`).

Hmm how would directives like this be translated then?

```c
#define UL(X) X ## UL
#define ONE UL(1)
```

File: bindgen-tests/tests/expectations/tests/macro_const_1_0.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply bindgen-tests/tests/expectations/tests/macro_const_1_0.rs:13]: hmmm interesting. I think it is fine as it is then.

**Comment by reitermarkus** [comment]
I think this might actually expand `ONE` to `1 as c_ulong` since there is special treatment for concatenating with a literal suffix like `UL`.

**Comment by reitermarkus** [comment]
Is should be working now, there was a bug that treated `1 ## U ## LL` different than `1 ## LL # U` which prevented some variants from being parsed.

**Comment by reitermarkus** [comment]
Nevermind. The `UL` macro is generated correctly as

```rust
macro_rules! __cmacro__UL {
  ($X:ident) => {
    concat_idents!($X, UL)
  };
}
```

`ONE` currently generates

```rust
pub const ONE: _ = concat_idents!(X , UL);
```

so some special handling in code generation needs to be done as well in addition to doing it in the parser.

**Comment by reitermarkus** [comment]
`ONE` should now correctly generate

```rust
pub const ONE: _ = 1;
```

File: bindgen-tests/tests/expectations/tests/macro_const.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: > they are treated as arrays first

But what's the difference? Aren't arrays just syntactic sugar for pointer arithmetic?

> What do you think?

I guess in the end using `&CStr` would probably be the most ergonomic, since it provides an `as_ptr` as well as `to_str` method. Currently this would require `unsafe` because only `[from_bytes_with_nul_unchecked](https://doc.rust-lang.org/core/ffi/struct.CStr.html#method.from_bytes_with_nul_unchecked)` is `const`-stable so far.

File: bindgen/callbacks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/callbacks.rs:72]: In practice you would only use `func_macro` for macros which cannot be parsed or cannot be generated by `cmacro`. Improving `cmacro` would then of course be the better alternative to using this function.

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/callbacks.rs:72]: Maybe we can just remove `func_macro` since `cmacro` supports function-like macros while `cexpr` didn't.

File: bindgen-tests/tests/expectations/tests/macro_const.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: Well arrays (or even just slices) intrinsically have more information because they can provide their size which is something that rust users could be relying on and we would be removing that part from the API by using `*const c_char`

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: I think my only concern is `bindgen` currently allows denying unsafe code, but I guess we can just use `&CStr` and annotate all strings with `#[allow(unsafe_code)]`.

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const.rs:17]: Okay, what should happen in this edge case?

```c
#define HELLO1 "WORLD"
#define HELLO2 (char* const)"WORLD"
```

Should `HELLO2` be a pointer because of the explicit cast or also a `&CStr`?

[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen-tests/tests/expectations/tests/macro_const.rs]: I have made both a `&CStr` now. That way you can call `as_ptr()` or `to_bytes_with_nul()` depending on your use case.

**Comment by pvdrz** [comment]
Hi @reitermarkus. I haven't had enough time to review the most recent changes but I suspect we need to figure out a way to avoid breaking the `--rust-target` flag with this feature as `cmacro` is able to emit arbitrary Rust code. I suggest two options depending on your priorities:

- We could put all these changes behind the `"experimental"` feature. In that way we can merge this faster and incrementally fix any regressions caused by the code generation of `cmacro`.
- Be sure that `cmacro` will emit valid code for the 1.30 version of the compiler (which is the oldest version we haven't deprecated).

In either case we need to be able to test that different versions of rustc can compile the code emitted by `bindgen` which means fixing https://github.com/rust-lang/rust-bindgen/issues/2277.

PD: I really want to merge this PR as it would fix a lot of weird issues we have with preprocessing directives.

**Comment by reitermarkus** [comment]
I guess we can also add some support for `--rust-target` in `cmacro` so we can either not generate code for older versions (e.g. for function-like macros) or alter code generation.

**Comment by reitermarkus** [comment]
I have now put the code generation for function-like macros behind the `experimental` feature.

Variable-like macros should already always generate the correct code.

**Comment by reitermarkus** [comment]
@pvdrz, please have another look here. Probably best to review https://github.com/rust-lang/rust-bindgen/pull/2543 first though.

**Comment by jschwe** [comment]
I'm wondering what the current status is here, is this PR blocked waiting on a review?

**Comment by pvdrz** [comment]
This PR is blocked (or superseeded) by https://github.com/rust-lang/rust-bindgen/pull/2561.

**Comment by alex** [comment]
@pvdrz Are you sure that's the right issue? On first glance it doesn't look the same.

**Comment by reitermarkus** [comment]
@alex, it's blocked by https://github.com/rust-lang/rust-bindgen/pull/2561 at the moment, i.e. part of this PR depends on the sorting being fixed.

**Comment by reitermarkus** [comment]
Note that this PR also includes the same changes.

**Comment by alex** [comment]
Ah, got it, I see how it's blocked. It was the superseded that confused me.

**Comment by reitermarkus** [comment]
So further testing/reviewing of https://github.com/rust-lang/rust-bindgen/pull/2561 would be appreciated to get this moving again.

File: bindgen/ir/context.rs [hunk 13]
```diff
@@ -2842,6 +3043,107 @@ If you encounter an error missing from this list, please file an issue or a PR!"
...
+            return Some((arg_types, ret_type));
+        }
+
+        None
+    }
+
+    fn rust_target(&self) -> Option<String> {
+        // FIXME: Workaround until `generate_cstr` is the default.
+        if !self.options().generate_cstr {
+            return Some("1.0".into());
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply bindgen/ir/context.rs]: Thanks, should be fixed now.

**Comment by overhacked** [comment]
@reitermarkus, this is really exciting because it would resolve [a downstream bug I ran into with `bssl-sys`](https://bugs.chromium.org/p/boringssl/issues/detail?id=636). The review comments above looked positive, any idea when it might be merged?

**Comment by reitermarkus** [comment]
@overhacked, thanks for testing. 

> any idea when it might be merged?

No idea when this will be merged, I already split some of the changes into separate PRs to make them easier to review, but all of them seem to also have stalled.

**Comment by therishidesai** [comment]
Any update from the bindgen team on when this will get merged?
